File: backend/app/spawn.py
"""Spawn logic: given a zone and date, roll a (possibly null) encounter."""
import random
from datetime import date

from . import events, species

# Default shiny rate is 1/64 — much more discoverable than the classic-Pokémon
# 1/512 / 1/4096, and calibrated for short play sessions in a single zone.
# Players can toggle a "shiny boost" in the settings menu to push it to 1/16
# for testing / casual play.
BASE_SHINY_RATE = 1 / 64
BOOST_SHINY_RATE = 1 / 16
DEFAULT_RATE = 0.05  # for species with no event gating

_shiny_boost = False


def get_shiny_boost() -> bool:
    return _shiny_boost


def set_shiny_boost(value: bool) -> None:
    global _shiny_boost
    _shiny_boost = bool(value)


def current_shiny_rate() -> float:
    return BOOST_SHINY_RATE if _shiny_boost else BASE_SHINY_RATE


def roll_encounter(zone: str, on: date, rng: random.Random | None = None) -> tuple[species.Species, bool] | None:
    """
    Returns (Species, is_shiny) if an encounter triggers, else None.

    Each candidate species in the zone contributes its current rate. If multiple species
    fire on the same step (rare), pick one weighted by rate.
    """
    rng = rng or random.Random()
    candidates: list[tuple[species.Species, float]] = []
    for s in species.species_for_zone(zone):
        rate = _rate_for(s, on)
        if rate <= 0:
            continue
        if rng.random() < rate:
            candidates.append((s, rate))

    if not candidates:
        return None

    # Weighted pick among species that triggered
    total = sum(r for _, r in candidates)
    pick = rng.uniform(0, total)
    acc = 0.0
    for s, r in candidates:
        acc += r
        if pick <= acc:
            chosen = s
            break
    else:
        chosen = candidates[-1][0]

    is_shiny = events.is_active(chosen.event, on) and rng.random() < current_shiny_rate()
    return chosen, is_shiny


def _rate_for(s: species.Species, on: date) -> float:
    if s.event is None:
        return DEFAULT_RATE
    e = events.get_event(s.event)
    if e is None:
        return DEFAULT_RATE
    # Use the override-aware is_active so the settings menu actually moves rates.
    return e.active_rate if events.is_active(s.event, on) else e.dormant_rate

File: frontend/src/game/spawn.ts
// Spawn logic — ported from backend/app/spawn.py.

import { getEvent, isActive } from "./data/events";
import { speciesForZone, type Species } from "./data/species";

export const BASE_SHINY_RATE = 1 / 64;
export const BOOST_SHINY_RATE = 1 / 16;
export const DEFAULT_RATE = 0.05;
export const ENCOUNTER_RATE_MULTIPLIER = 0.5;

let _shinyBoost = false;

export function getShinyBoost(): boolean { return _shinyBoost; }
export function setShinyBoost(value: boolean): void { _shinyBoost = !!value; }
export function currentShinyRate(): number {
  return _shinyBoost ? BOOST_SHINY_RATE : BASE_SHINY_RATE;
}

function rateFor(s: Species, isoDate: string): number {
  if (s.event === null) return DEFAULT_RATE * ENCOUNTER_RATE_MULTIPLIER;
  const e = getEvent(s.event);
  if (!e) return DEFAULT_RATE * ENCOUNTER_RATE_MULTIPLIER;
  const base = isActive(s.event, isoDate) ? e.active_rate : e.dormant_rate;
  return base * ENCOUNTER_RATE_MULTIPLIER;
}

/** Returns [Species, isShiny] if an encounter triggers, else null. */
export function rollEncounter(zone: string, isoDate: string): [Species, boolean] | null {
  const candidates: Array<[Species, number]> = [];
  for (const s of speciesForZone(zone)) {
    const rate = rateFor(s, isoDate);
    if (rate <= 0) continue;
    if (Math.random() < rate) candidates.push([s, rate]);
  }
  if (candidates.length === 0) return null;

  const total = candidates.reduce((a, [, r]) => a + r, 0);
  let pick = Math.random() * total;
  let chosen = candidates[candidates.length - 1][0];
  for (const [s, r] of candidates) {
    pick -= r;
    if (pick <= 0) { chosen = s; break; }
  }

  const shiny = isActive(chosen.event, isoDate) && Math.random() < currentShinyRate();
  return [chosen, shiny];
}
